File: lib/api/models/ai_feedback.dart
// lib/models/ai_feedback.dart

class AIFeedback {
  final int? feedbackId;
  final int interactionId;
  final int userId;
  final int? rating;
  final String? feedbackText;
  final DateTime? createdAt;
  final bool isUsedForImprovement;

  // Relationships
  final Map<String, dynamic>? interaction;
  final Map<String, dynamic>? user;

  AIFeedback({
    this.feedbackId,
    required this.interactionId,
    required this.userId,
    this.rating,
    this.feedbackText,
    this.createdAt,
    this.isUsedForImprovement = false,
    this.interaction,
    this.user,
  });

  factory AIFeedback.fromJson(Map<String, dynamic> json) {
    return AIFeedback(
      feedbackId: json['feedback_id'],
      interactionId: json['interaction_id'],
      userId: json['user_id'],
      rating: json['rating'],
      feedbackText: json['feedback_text'],
      createdAt: json['created_at'] != null
          ? DateTime.parse(json['created_at'])
          : null,
      isUsedForImprovement: json['is_used_for_improvement'] ?? false,
      interaction: json['interaction'] != null
          ? Map<String, dynamic>.from(json['interaction'])
          : null,
      user:
          json['user'] != null ? Map<String, dynamic>.from(json['user']) : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      if (feedbackId != null) 'feedback_id': feedbackId,
      'interaction_id': interactionId,
      'user_id': userId,
      if (rating != null) 'rating': rating,
      if (feedbackText != null) 'feedback_text': feedbackText,
      if (createdAt != null) 'created_at': createdAt!.toIso8601String(),
      'is_used_for_improvement': isUsedForImprovement,
      if (interaction != null) 'interaction': interaction,
      if (user != null) 'user': user,
    };
  }

  AIFeedback copyWith({
    int? feedbackId,
    int? interactionId,
    int? userId,
    int? rating,
    String? feedbackText,
    DateTime? createdAt,
    bool? isUsedForImprovement,
    Map<String, dynamic>? interaction,
    Map<String, dynamic>? user,
  }) {
    return AIFeedback(
      feedbackId: feedbackId ?? this.feedbackId,
      interactionId: interactionId ?? this.interactionId,
      userId: userId ?? this.userId,
      rating: rating ?? this.rating,
      feedbackText: feedbackText ?? this.feedbackText,
      createdAt: createdAt ?? this.createdAt,
      isUsedForImprovement: isUsedForImprovement ?? this.isUsedForImprovement,
      interaction: interaction ?? this.interaction,
      user: user ?? this.user,
    );
  }
}

File: final_project/models/ai_feedback.py
from sqlalchemy import Column, Integer, Text, Boolean, DateTime, CheckConstraint, ForeignKey, Index
from sqlalchemy.sql import func
from sqlalchemy.orm import relationship
from final_project.db import Base

# SQLAlchemy model for storing AI feedback from users
class AIFeedbackDB(Base):
    __tablename__ = "ai_feedback"

    feedback_id = Column(Integer, primary_key=True, autoincrement=True)  # Primary key for feedback
    interaction_id = Column(Integer, ForeignKey("ai_interaction.interaction_id"), nullable=False)  # Foreign key to AI interaction
    user_id = Column(Integer, ForeignKey("users.user_id"), nullable=False)  # Foreign key to user providing feedback
    rating = Column(Integer)  # Rating from 1 to 5
    feedback_text = Column(Text)  # Optional feedback text
    created_at = Column(DateTime, server_default=func.now())  # Timestamp when feedback was created
    is_used_for_improvement = Column(Boolean, default=False)  # Flag to indicate if feedback was used for AI improvement

    # Constraints and indexes for performance and data integrity
    __table_args__ = (
        CheckConstraint("rating >= 1 AND rating <= 5", name="check_rating_range"),  # Ensure rating is between 1 and 5
        Index("idx_ai_feedback_interaction_id", "interaction_id"),  # Index on interaction_id
        Index("idx_ai_feedback_user_id", "user_id"),  # Index on user_id
        Index("idx_ai_feedback_rating", "rating"),  # Index on rating
        Index("idx_ai_feedback_is_used_for_improvement", "is_used_for_improvement"),  # Index on is_used_for_improvement
    )

    # Define relationships
    interaction = relationship("AIInteractionDB", backref="feedback")  # Relationship with AIInteractionDB
    user = relationship("UserDB", backref="ai_feedback")  # Relationship with UserDB
